contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,325 @@
+
+This maintained self-hosted users’ freedom, while still enabling Cloud users to benefit from community apps. Apps are all developed and tested in the open. We just need to do a final manual check before installing them. After launch, we found many of [our most used apps](/blog/the-state-of-plugins) came from our community, proving this decision’s success.
+
+### What did we learn from all this?
+
+- Python, which we wrote our ingestion pipeline in, wasn’t going to work for apps. Dependency management for arbitrary code was too difficult.
+
+- Apps needed to scale while having access to some key JavaScript libraries. Using Node and Celery enabled us to integrate into our ingestion pipeline while providing the functionality we needed.
+
+- We needed to make sure the code apps were running was secure and sandboxed. VMs were the answer to this, and Node had support for VMs.
+
+- The default NodeVM wasn’t secure and couldn’t communicate with a host. `isolated-vm` was secure but needed proxying. `vm2` had the ability to proxy code, but was less secure. `vm2` and a review process for Cloud apps was the ultimate decision.
+
+These learnings led to our current app structure.
+
+## 📈 Part 3: Building a mature app platform that scales
+
+After creating a basic MVP, and seeing the benefits of its usage, work continued to add functionality and support greater scale. 
+
+Before we get any deeper, it's worth recapping the structure and features of apps as they are now. 
+
+### App structure
+
+The basic structure of an app has stayed the same since the JavaScript rebuild. Each app contains two key files:
+
+- `index.js` (or `index.ts`) which contains key application logic, modifies events, connects with external services, and interacts with PostHog. It has an allowlist of libraries to use for utilities or interacting with other services. It also has access (through the VM) to meta information, storage, and logging.
+
+- `plugin.json` contains configuration data for the app, what shows up in PostHog, and variables for end users to modify. The file contains the name, description, and other details that show up on the front end.
+
+Here’s what a basic `index.js` looks like:
+
+```js
+// Some internal library function
+async function getRandomNumber() {
+    return 4
+}
+
+// Plugin method that runs on plugin load
+async function setupPlugin({ config }) {
+    console.log(config.greeting)
+}
+
+// Plugin method that processes event
+async function processEvent(event, { config, cache }) {
+    const counterValue = (await cache.get('greeting_counter', 0))
+    cache.set('greeting_counter', counterValue + 1)
+    if (!event.properties) event.properties = {}
+    event.properties['greeting'] = config.greeting
+    event.properties['greeting_counter'] = counterValue
+    event.properties['random_number'] = await getRandomNumber()
+    return event
+}
+
+// The plugin itself
+module.exports = {
+    setupPlugin,
+    processEvent
+}
+```
```

Comment:
```suggestion
```js
// Some internal library function
async function getRandomNumber() {
    return 4
}

// Plugin method that runs on plugin load
export async function setupPlugin({ config }) {
    console.log(config.greeting)
}

// Plugin method that processes event
export async function processEvent(event, { config, cache }) {
    const counterValue = (await cache.get('greeting_counter', 0))
    cache.set('greeting_counter', counterValue + 1)
    if (!event.properties) event.properties = {}
    event.properties['greeting'] = config.greeting
    event.properties['greeting_counter'] = counterValue
    event.properties['random_number'] = await getRandomNumber()
    return event
}
```
```

Changed `module.exports` to `export` in front of both `setupPlugin` and `processEvent`